src/webviews/processor.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { existsSync, readFileSync } from 'node:fs';
+import { Uri, ViewColumn, WebviewPanel, env, window } from 'vscode';
+
+export const MESSAGING_JS_PATH = 'resources/instructions/webviewMessaging.js';
+export const MESSAGING_SCRIPT_PATH_DEMARCATOR = '--- MESSAGING_SCRIPT_SRC ---';
+
+export type WebviewMessageCallback = (responseData?: object) => void;
+
+export type WebviewMessageHandler = {
+    type: string;
+    action: (
+        panel: WebviewPanel,
+        data?: object,
+        callback?: WebviewMessageCallback
+    ) => void;
+};
+
+export class WebviewProcessor {
+    extensionUri: Uri;
+
+    constructor(extensionUri: Uri) {
+        this.extensionUri = extensionUri;
+    }
+
+    public createWebviewPanel(viewType: string, title: string): WebviewPanel {
+        const panel = window.createWebviewPanel(
+            viewType,
+            title,
+            ViewColumn.Beside,
+            {
+                enableScripts: true,
+                localResourceRoots: [this.extensionUri]
+            }
+        );
+        return panel;
+    }
+
+    public static validateMessageHanders(
+        messageHandlers: WebviewMessageHandler[]
+    ) {
+        const handlerMap: { [type: string]: boolean } = {};
+        for (const handler of messageHandlers) {
+            if (handlerMap[handler.type] === true) {
+                throw new Error(
+                    `There can be only one message handler per type. There are at least two handlers with type '${handler.type}'.`
+                );
+            } else {
+                handlerMap[handler.type] = true;
+            }
+        }
+    }
+
+    public onWebviewReceivedMessage(
+        data: any,
+        panel: WebviewPanel,
+        messageHandlers: WebviewMessageHandler[]
+    ) {
+        const responsiveHandlers = messageHandlers.filter(
+            (messageHandler) => data.type === messageHandler.type
+        );
+        if (responsiveHandlers.length > 0) {
+            const handler = responsiveHandlers[0];
```

Comment:
nit: Refactor into a new file(processor.ts) hides the assertion that why index 0 handler is a valid handler. Since the validation of handler map takes place at all the caller(`showInstructionWebview`) maybe `onWebviewReceivedMessage` should be passed a single message handler instead of an array there?

Reply:
That's a good catch! In my next PR I'm actually moving the handler validation into the processor as well, for the reason you stated and the reason that it's really more congruent there (another aspect you pointed out).

While I like having the array just passed in here for its own processing, I do agree that the selection of just the first item feels a little oblique. I could either add a comment here, or what I'm leaning toward is just switching to `messageHandlers.find()` instead, which only returns the first item found (which is always going to be 0 or 1 items), and probably reads better for this use case. Let me know what you think.